File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_13.cpp
// 1
// 2 3
// 4 5 6
// 7 8 9 10
// 11 12 13 14 15

//----------------------------------
// BRUTE FORCE APPROACH
//----------------------------------
#include <iostream>
using namespace std;
int main()
{
    int n = 5;
    int num = 1;
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= i; j++)
        {
            cout << num++ << " ";
        }
        cout << endl;
    }
    return 0;
}

//----------------------------------
// BETTER APPROACH (Using string building)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    int num = 1;
    for (int i = 1; i <= n; i++)
    {
        string s = "";
        for (int j = 1; j <= i; j++)
        {
            s += to_string(num++) + " ";
        }
        cout << s << endl;
    }
    return 0;
}

//----------------------------------
// BEST APPROACH (Correct + Efficient)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;

int main()
{
    int n = 5;
    int num = 1;
    // Rough pre-allocation to reduce reallocations
    string s;
    s.reserve(n * n * 3); // safe guess, avoids repeated memory resizing

    for (int i = 1; i <= n; i++)
    {
        s = ""; // reset string for each row
        for (int j = 1; j <= i; j++)
        {
            s += to_string(num++) + " ";
        }
        cout << s << endl;
    }
    return 0;
}

File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_13.py
# 1
# 2 3
# 4 5 6
# 7 8 9 10
# 11 12 13 14 15

#----------------------------------
# BRUTE FORCE APPROACH
#----------------------------------
n = 5
num = 1

for i in range(1, n + 1):
    for j in range(1, i + 1):
        print(num, end=" ")
        num += 1
    print()


#----------------------------------
# BETTER APPROACH (Using string building)
#----------------------------------
n = 5
num = 1

for i in range(1, n + 1):
    s = ""
    for j in range(1, i + 1):
        s += str(num) + " "
        num += 1
    print(s)


#----------------------------------
# BEST APPROACH (Correct + Efficient)
#----------------------------------
n = 5
num = 1

for i in range(1, n + 1):
    row = " ".join(str(num + k) for k in range(i))
    num += i
    print(row)
